### ISSUE TITLE
ssl.SSLError: unknown error with bundle_files=0

### ISSUE THREAD

**Issue by neodyne** [py2exe/py2exe#65]
Briefly mentioned in #63.

ugly example code, Main.py:

```
import json
from urllib import request

w=json.loads(request.urlopen('https://httpbin.org/headers').read().decode('utf-8'))['headers']['User-Agent']
print(w)
```
setup.py:

```
from distutils.core import setup
import py2exe

setup(console=[{ "script": "Main.py"}],
    options={"py2exe": {
        "bundle_files": 0}},
    zipfile=None)
```
Instead of the expected `'Python-urllib/3.8'` I get the following error:

```
Traceback (most recent call last):
  File "Main.py", line 4, in <module>
    w=json.loads(request.urlopen('https://httpbin.org/headers').read().decode('utf-8'))['headers']['User-Agent']
  File "urllib\request.pyc", line 222, in urlopen
  File "urllib\request.pyc", line 525, in open
  File "urllib\request.pyc", line 542, in _open
  File "urllib\request.pyc", line 502, in _call_chain
  File "urllib\request.pyc", line 1393, in https_open
  File "urllib\request.pyc", line 1319, in do_open
  File "http\client.pyc", line 1394, in __init__
  File "ssl.pyc", line 737, in create_default_context
  File "ssl.pyc", line 483, in __new__
ssl.SSLError: unknown error (_ssl.c:3152)
```
In python 3.7 and up, _ssl.pyd relies on libcrypto and libssl, as pointed out [here](https://github.com/py2exe/py2exe/pull/13). py2exe (on python 3.8.7) copies
```
libcrypto-1_1.dll
libssl-1_1.dll
libffi-7.dll
```
from Python's `DLLs` subdirectory to the build directory when using `bundle_files>=1`, but these files are missing from the zipfile with `bundle_files=0`.

**Comment by albertosottile** [comment]
I will look into this as soon as I have some spare time. In the meantime, could you modify your `setup.py` script to include a `"verbose": 4` entry in your `py2exe` dictionary (see e.g. [here](https://github.com/py2exe/py2exe/blob/bundle_files/tests/functional/bundlefiles0_test/setup.py)) and paste here the log obtained while running the setup script? Thanks.

**Comment by neodyne** [comment]
The log:
```
running py2exe

  3 missing Modules
  ------------------
? _posixshmem                imported from multiprocessing.resource_tracker, multiprocessing.shared_memory
? readline                         imported from cmd, code, pdb
? test                                imported from multiprocessing.util
Building 'dist\Main.exe'.
Add resource PYTHONSCRIPT/1(2268 bytes) to dist\Main.exe
Add resource python38.dll/1(4207672 bytes) to dist\Main.exe
Add _decimal.pyd to dist\Main.exe
Add _bz2.pyd to dist\Main.exe
Add _lzma.pyd to dist\Main.exe
Add _asyncio.pyd to dist\Main.exe
Add _hashlib.pyd to dist\Main.exe
Add _ctypes.pyd to dist\Main.exe
Add _ssl.pyd to dist\Main.exe
Add _overlapped.pyd to dist\Main.exe
Add _multiprocessing.pyd to dist\Main.exe
Add select.pyd to dist\Main.exe
Add _queue.pyd to dist\Main.exe
Add pyexpat.pyd to dist\Main.exe
Add _socket.pyd to dist\Main.exe
Add unicodedata.pyd to dist\Main.exe
Add DLL libffi-7.dll to dist\Main.exe
Add DLL libssl-1_1.dll to dist\Main.exe
Add DLL libcrypto-1_1.dll to dist\Main.exe
```
So it seems that the DLLs get added after all. Ditching `zipfile=None` and going carefully through the contents of `dist/library.zip`, they are there as well. I apologize, don't know how I missed them the first time around; I changed the issue name to properly reflect the problem. Still, the error when run remains the same.

**Comment by albertosottile** [comment]
The issue is indeed there, and this time I truly have no idea on how to fix it: more than a bug, this seems like the lack of a feature. I do not have proof of this but, looking at the code of py2exe I am 99% sure that importing arbitrary DLLs from the executable/zipfile (so, with `bundle_files=0`) has never worked.

On paper, `MemoryModule` should work with arbitrary DLL files, not just with Python extensopms (PYD). In reality, all the import stage is handled by `zipextimporter` which does not contain the code to do so, as it inheriting from `zipimport`, which is a Python module designed to import other Python modules or packages from a zip archive. 

In addition to that, nothing in the Python structure is aware of the external DLL dependencies of the bundled extensions: this is typically handled by the linking of the libraries themselves, which triggers the system calls that attempt to load the DLLs from the file system (and fail because the file is not there). To make somehow this work, one would have to intercept the calls to `LoadLibrary` from the PYD files and replace them with calls to `MyLoadLibrary` which *could* load the DLL from the archive and supply it to the calling PYDs. I have no idea if this could work or not but, for sure, none of this code is currently in py2exe.

Hence, after this investigation, I am convinced that `bundle_files=0` only works when **no external DLL** is needed for the software to function. If anyone can find an example case from py2exe 0.9.x that actually contradicts this assumption, I would be more than glad to run it and to dissect to see how could that work. 

So, the reason why your example works with 3.6 and lower is that up to then there was no need for the `_ssl` module to load external libraries. From 3.7 and onwards, `_ssl` depends on `libssl-1_1.dll` and `libcrypto-1_1.dll`, and py2exe is just unable to load them from the zipfile (or the executable, fwiw).

In summary, I am not sure how to proceed with this issue. For the moment I am marking it as ehnancement/help wanted, but I do not see a clear path forward for implementing this feature. Once again, if anyone has suggestions or ideas on this topic, comments and PRs are always welcome.

(Side note: I could not run your example in Docker because of
```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1091)
```
so I resorted in using the SSL test already available in the test folder).

**Comment by albertosottile** [comment]
Just found this https://github.com/py2exe/py2exe/blob/93c4d9ea21c0b3da86b5008fb2f7785b7934af81/source/_memimporter.c#L179-L182

So, maybe the feature was there a long time ago and was then stripped because it was incompatible with too many libraries.

**Comment by neodyne** [comment]
It may be of note that the comment is correct even with the **latest** py2exe version. `sqlite3.dll` is indeed added to the zip from `Python\DLLs` and properly loaded, so at least some DLLs _can_ be loaded from the zipfile. This can be tested with `bundle_files=0` and an example script:
```
import sqlite3
print(sqlite3)
```
which runs without issues. The example is included in the original [README](https://github.com/py2exe/py2exe/blob/master/README_ORIGINAL.rst#the-bundle-files-option-explained) for the python 3 rewrite. Maybe special case code that allows this exists, which could now be extended to include the ssl libs? After all, `ssl` is not a third-party module, unlike `numpy`.

**Comment by albertosottile** [comment]
You are absolutely right. This demonstrates even further that I do not know the codebase of `py2exe` well enough... :/ 

Anyway, after long sessions with WinDbg, it is clear to me now that the libraries are correctly loaded (`_AddMemoryModule` and `_FindMemoryModule` do part of the magic that I was describing in the comment above):

```
  _AddMemoryModule(libcrypto-1_1.dll, 000001EFE46D82A0) -> 000001EFE471EF90[1]
  _LoadLibrary(libcrypto-1_1.dll, 000001EFE6404580) -> 000001EFE46D82A0 libcrypto-1_1.dll[1]
  _FindMemoryModule((null), 000001EFE46D82A0) -> libcrypto-1_1.dll[1]
[...]
 _AddMemoryModule(libssl-1_1.dll, 000001EFE4688B50) -> 000001EFE4688C40[1]
  _LoadLibrary(libssl-1_1.dll, 000001EFE6404580) -> 000001EFE4688B50 libssl-1_1.dll[1]
  _FindMemoryModule((null), 000001EFE4688B50) -> libssl-1_1.dll[1]
[...]
  _AddMemoryModule(_ssl.pyd, 000001EFE4735AE0) -> 000001EFE6825310[1]
_LoadLibrary(_ssl.pyd, 000001EFE6404580) -> 000001EFE4735AE0 _ssl.pyd[1]
_FindMemoryModule((null), 000001EFE4735AE0) -> _ssl.pyd[1]
```
Furthermore, if I remove one of the DLLs, I get a very different error from `_memimporter` (`The specified module could not be found`).

Nevertheless, even if the libraries are loaded well, `_ssl.c` still does not manage to correctly create the context. To me it looks quite clear that the issue is in `cpython/Modules/_ssl.c` https://github.com/python/cpython/blob/f08c66467d56c71f75c6659ede9177449fe9d2e6/Modules/_ssl.c#L3091 or, most likely, in `openssl/ssl/ssl_lib.c`  https://github.com/openssl/openssl/blob/e2e09d9fba1187f8d6aafaa34d4172f56f1ffb72/ssl/ssl_lib.c#L2961

> After all, ssl is not a third-party module, unlike numpy.

I tend to agree, in principle. However, when the external libraries are correctly loaded yet do not function when loaded from memory (gently reminder that they should not, as the regular loading mechanism of Windows is from the file system), there's not much more I can do. I probably do not have enough knowledge on this topic but, I am not aware of a way to fix this via changing te `py2exe` codebase.

Once again, I do not see a clear path forward for fixing this issue from py2exe. If anyone has suggestions or ideas on this topic, comments and PRs are always welcome.

**Comment by albertosottile** [comment]
TIL that also PyOxidizer uses `MemoryModule` to load DLLs from memory, it would be interesting to know how that tools works when embedding `_ssl`.

**Comment by neodyne** [comment]
I'm not sure that the general terminating error when using `ssl` with `bundle_files=0` is a good idea, in the latest released build. It might make sense for the sake of identical behavior across all supported python versions, but, in my opinion, the build should hard fail only on python 3.7+, since the issue is not present on 3.6.

**Comment by albertosottile** [comment]
>  the build should hard fail only on python 3.7+

That's already the way it is implemented:
https://github.com/py2exe/py2exe/blob/d002d4ea7f49e3710dc37bf5b90432be33ce4b09/py2exe/hooks.py#L643-L653

**Comment by albertosottile** [comment]
This is fixed in 0.10.2.1.
